### PR TITLE
scripted-diff: TxoutType C++11 scoped enum class

### PR DESCRIPTION
**Pull request by maflcko**
Non-scoped enums can accidentally and silently decay into an integral type. Also, the symbol names of the keys are exported to the surrounding (usually global) namespace.

Fix both issues by switching to an `enum class TxoutType` in a (mostly) scripted-diff.

### CODE REVIEW

**Comment by practicalswift** [comment]
Concept ACK: [Prefer class enums over “plain” enums](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Renum-class) (C++ Core Guidelines)

These are the remaining non-class enums that might be worth investigating too:

```
$ git grep '^enum [^c][^l][^(]*$' -- "*.h" "*.cpp" ":(exclude)src/univalue/" 
      ":(exclude)src/leveldb/" | cut -f2 -d' ' | cut -f1 -d: | sort -u | tr "\n" " "
BindFlags ChangeType DeploymentPos DisconnectResult GetDataMsg HTTPStatusCode
isminetype NetPermissionFlags Network NumConnections opcodetype RPCErrorCode SafeChars
ServiceFlags SOCKS5Atyp SOCKS5Command SOCKS5Method SOCKS5Reply SOCKSVersion txnouttype
WalletFeature WalletFlags
```

**Comment by hebasto** [comment]
Concept ACK.
Also: #17877

File: src/script/standard.h [hunk 1]
```diff
@@ -36,11 +36,11 @@ static const unsigned int MAX_OP_RETURN_RELAY = 83;
 
 /**
  * A data carrying output is an unspendable output containing data. The script
- * type is designated as TX_NULL_DATA.
+ * type is designated as NULL_DATA.
```
Comment (hebasto) [inline src/script/standard.h:39]: nit:
```suggestion
 * type is designated as TxoutType::NULL_DATA.
```

File: src/script/standard.h [hunk 2]
```diff
@@ -36,11 +36,11 @@ static const unsigned int MAX_OP_RETURN_RELAY = 83;
...
  */
 extern bool fAcceptDatacarrier;
 
-/** Maximum size of TX_NULL_DATA scripts that this node considers standard. */
+/** Maximum size of NULL_DATA scripts that this node considers standard. */
```
Comment (hebasto) [inline src/script/standard.h:43]: nit:
```suggestion
/** Maximum size of TxoutType::NULL_DATA scripts that this node considers standard. */
```

File: src/script/standard.h [hunk 3]
```diff
@@ -157,9 +156,9 @@ std::string GetTxnOutputType(txnouttype t);
  *
  * @param[in]   scriptPubKey   Script to parse
  * @param[out]  vSolutionsRet  Vector of parsed pubkeys and hashes
- * @return                     The script type. TX_NONSTANDARD represents a failed solve.
+ * @return                     The script type. NONSTANDARD represents a failed solve.
```
Comment (hebasto) [inline src/script/standard.h:159]: nit:
```suggestion
 * @return                     The script type. TxoutType::NONSTANDARD represents a failed solve.
```

File: src/script/standard.cpp [hunk 4]
```diff
@@ -27,19 +27,20 @@ WitnessV0ScriptHash::WitnessV0ScriptHash(const CScript& in)
...
+    case TxoutType::PUBKEY: return "pubkey";
+    case TxoutType::PUBKEYHASH: return "pubkeyhash";
+    case TxoutType::SCRIPTHASH: return "scripthash";
+    case TxoutType::MULTISIG: return "multisig";
+    case TxoutType::NULL_DATA: return "nulldata";
+    case TxoutType::WITNESS_V0_KEYHASH: return "witness_v0_keyhash";
+    case TxoutType::WITNESS_V0_SCRIPTHASH: return "witness_v0_scripthash";
+    case TxoutType::WITNESS_UNKNOWN: return "witness_unknown";
+        // no default case, so the compiler can warn about missing cases
     }
```
Comment (hebasto) [inline src/script/standard.cpp:44]: nit:
```suggestion
    } // no default case, so the compiler can warn about missing cases
```
as described in [Developer Notes](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#c-data-structures)

**Comment by MarcoFalke** [comment]
@practicalswift It might be best to open a new brainstorming issue for discussion about those. I think enum flags can not (or should not?) be switched to `enum class`.

**Review by hebasto — COMMENTED** [review]
ACK fa38a7472ab475806cf50f302db9c69dfb8ac0f5, tested on Linux Mint 19.3 (x86_64).

fa868d3a8de56582538a3e4ef8de6082f2b8c747:
maybe make a separate utility function to derive an underlying type of an enumerator?

File: src/script/standard.h (hunk 1 above)
Comment (MarcoFalke) [reply src/script/standard.h:39]: I think in comments we don't need to be overly verbose. Especially not in the file that defines these values. I think the comment is clearer as is.

File: src/script/standard.h (hunk 2 above)
Comment (MarcoFalke) [reply src/script/standard.h:43]: I think in comments we don't need to be overly verbose. Especially not in the file that defines these values. I think the comment is clearer as is.

File: src/script/standard.h (hunk 3 above)
Comment (MarcoFalke) [reply src/script/standard.h:159]: I think in comments we don't need to be overly verbose. Especially not in the file that defines these values and where the return type is mentioned one line below the comment. I think the comment is clearer as is.

File: src/script/standard.cpp (hunk 4 above)
Comment (MarcoFalke) [reply src/script/standard.cpp]: Thanks. Done

**Review by hebasto — APPROVED** [review]
ACK fa5997bd6fc82e16b597ea96e3c5c665f1f174ab, since fa38a7472ab475806cf50f302db9c69dfb8ac0f5 suggested changes applied.

**Comment by MarcoFalke** [comment]
Rebased due to merge conflict in adjacent line. Can be trivially re-ACKed with git range-diff

**Comment by practicalswift** [comment]
ACK fa32adf9dc25540ad27f5b82654c7057d7738627 -- patch looks correct

**Review by hebasto — APPROVED** [review]
re-ACK fa32adf9dc25540ad27f5b82654c7057d7738627, since fa5997bd6fc82e16b597ea96e3c5c665f1f174ab (https://github.com/bitcoin/bitcoin/pull/19114#pullrequestreview-421425198) rebased only (verified with `git range-diff`).

**Comment by hebasto** [comment]
@MarcoFalke from which key-server I could receive your non-expired key to verify commit signatures?

**Comment by MarcoFalke** [comment]
@hebasto See #18385

**Comment by hebasto** [comment]
@MarcoFalke 
> @hebasto See #18385

Thanks. It was helpful!

**Pull request merged**